rust/src/lib.rs

```diff
@@ -106,430 +139,387 @@ pub extern "C" fn zcashlc_clear_last_error() {
+///   of pointer::offset.
+/// - Call [`zcashlc_string_free`] to free the memory associated with the returned pointer
+///   when you are done using it.
 #[no_mangle]
-pub unsafe extern "C" fn zcashlc_derive_extended_full_viewing_key(
-    extsk: *const c_char,
+pub extern "C" fn zcashlc_spending_key_to_full_viewing_key(
+    usk_ptr: *const u8,
+    usk_len: usize,
     network_id: u32,
 ) -> *mut c_char {
     let res = catch_panic(|| {
         let network = parse_network(network_id)?;
-        let extsk = CStr::from_ptr(extsk).to_str()?;
-        let extfvk = match decode_extended_spending_key(
-            network.hrp_sapling_extended_spending_key(),
-            &extsk,
-        ) {
-            Ok(Some(extsk)) => ExtendedFullViewingKey::from(&extsk),
-            Ok(None) => {
-                return Err(format_err!("Deriving viewing key from spending key returned no results. Encoding was valid but type was incorrect."));
-            }
-            Err(e) => {
-                return Err(format_err!(
-                    "Error while deriving viewing key from spending key: {}",
-                    e
-                ));
-            }
-        };
-
-        let encoded = encode_extended_full_viewing_key(
-            network.hrp_sapling_extended_full_viewing_key(),
-            &extfvk,
-        );
-
-        Ok(CString::new(encoded).unwrap().into_raw())
+        unsafe { decode_usk(usk_ptr, usk_len) }.map(|usk| {
+            let ufvk = usk.to_unified_full_viewing_key();
+            CString::new(ufvk.encode(&network)).unwrap().into_raw()
+        })
     });
     unwrap_exc_or_null(res)
 }
 
-/// Initialises the data database with the given block.
+/// Initialises the data database with the given block metadata.
 ///
 /// This enables a newly-created database to be immediately-usable, without needing to
 /// synchronise historic blocks.
+///
+/// # Safety
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must have an
+///   alignment of `1`. Its contents must be a string representing a valid system path in the
+///   operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - `hash_hex` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `hash_hex` must not be mutated for the duration of the function call.
+/// - `sapling_tree_hex` must be non-null and must point to a null-terminated UTF-8 string
+///   containing the encoded byte representation of a Sapling commitment tree.
+/// - The memory referenced by `sapling_tree_hex` must not be mutated for the duration of the
+///   function call.
 #[no_mangle]
 pub extern "C" fn zcashlc_init_blocks_table(
```

Comment:
```suggestion
pub unsafe extern "C" fn zcashlc_init_blocks_table(
```

Reply:
Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.